src/pages/signin/ValidateCodeForm.js

```diff
@@ -0,0 +1,230 @@
+    constructor(props) {
+        super(props);
+        this.validateAndSubmitForm = this.validateAndSubmitForm.bind(this);
+        this.resendValidateCode = this.resendValidateCode.bind(this);
+        this.clearSignInData = this.clearSignInData.bind(this);
+
+        this.state = {
+            formError: false,
+            validateCode: '',
+            twoFactorAuthCode: '',
+        };
+    }
+
+    componentDidMount() {
+        if (!canFocusInputOnScreenFocus() || !this.inputValidateCode || !this.props.isVisible) {
+            return;
+        }
+        this.inputValidateCode.focus();
+    }
+
+    componentDidUpdate(prevProps, prevState) {
+        if (!prevProps.isVisible && this.props.isVisible) {
+            this.inputValidateCode.focus();
+        }
+        if (prevProps.isVisible && !this.props.isVisible && this.state.validateCode) {
+            this.clearValidateCode();
+        }
+        if (!prevProps.account.requiresTwoFactorAuth && this.props.account.requiresTwoFactorAuth) {
+            this.input2FA.focus();
+        }
+        if (prevState.twoFactorAuthCode !== this.state.twoFactorAuthCode && this.state.twoFactorAuthCode.length === CONST.TFA_CODE_LENGTH) {
+            this.validateAndSubmitForm();
+        }
+    }
+
+    /**
+     * Clear Validate Code from the state
+     */
+    clearValidateCode() {
+        this.setState({validateCode: ''}, this.inputValidateCode.clear);
+    }
+
+    /**
+     * Trigger the reset validate code flow and ensure the 2FA input field is reset to avoid it being permanently hidden
+     */
+    resendValidateCode() {
+        if (this.input2FA) {
+            this.setState({twoFactorAuthCode: ''}, this.input2FA.clear);
+        }
+        this.setState({formError: false});
+        User.resendValidateCode(this.props.credentials.login, true);
+    }
+
+    /**
+    * Clears local and Onyx sign in states
+    */
+    clearSignInData() {
+        this.setState({twoFactorAuthCode: '', formError: false});
+        Session.clearSignInData();
+    }
+
+    /**
+     * Check that all the form fields are valid, then trigger the submit callback
+     */
+    validateAndSubmitForm() {
+        if (!this.state.validateCode.trim()) {
+            this.setState({formError: 'validateCodeForm.error.pleaseFillMagicCode'});
+            return;
+        }
+
+        if (!ValidationUtils.isValidValidateCode(this.state.validateCode)) {
+            this.setState({formError: 'validateCodeForm.error.incorrectMagicCode'});
+            return;
+        }
+
+        if (this.props.account.requiresTwoFactorAuth && !this.state.twoFactorAuthCode.trim()) {
+            this.setState({formError: 'validateCodeForm.error.pleaseFillTwoFactorAuth'});
+            return;
+        }
+
+        this.setState({
+            formError: null,
+        });
+
+        Session.signIn('', this.state.validateCode, this.state.twoFactorAuthCode);
+    }
+
+    render() {
+        return (
+            <>
```

Comment:
The react fragment wrapper here caused the following issue:
- https://github.com/Expensify/App/issues/32683

Which is a Safari specific autofill popup issue that was observed by others, see this [SO (comment)](https://stackoverflow.com/questions/38663578/how-to-hide-autofill-safari-icon-in-input-field/57678169#57678169) and https://github.com/Expensify/App/issues/32683#issuecomment-2215352377 breakdown for more details.